PARQUET-212: Implement LIST read compatibility rules in Thrift

rdblue opened the pull request:
This implements the read-side compatibility rules for 2-level and 3-level lists in Thrift.

Thrift doesn't allow null elements inside lists, but 3-level lists may have optional elements. This PR adds a property, parquet.thrift.ignore-null-elements, that allows thrift to read lists with optional elements by ignoring nulls. This is off by default, but is provided as an opt-in for compatibility with data written by Hive.

Thrift's schema conversion does not change because a Thrift class (or Scrooge etc.) must be set in a file's metadata or provided when constructing a reader.

This replaces and closes #144.

rdblue commented:
@isnotinvain you may want to have a look at this since you guys do the most Thrift work. This adds support in Thrift to read data written by the other object models correctly.

isnotinvain commented:
I'll try to take a look thanks

rdblue commented:
Thanks, Alex!

@liancheng may have some comments as well since he is familiar with the compatibility rules.

liancheng commented on parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java:
```diff
@@ -753,13 +755,14 @@ static boolean isElementType(Type repeatedType, Schema elementSchema) {
       // The repeated type must be the element type because it matches the
       // structure of the Avro element's schema.
-      return true;
+      return fieldNames.contains(repeatedType.asGroupType().getFieldName(0));
```
I don't quite get the purpose of this change. Is this mainly for more robust support for schema evolution? Take the following Avro schema and Parquet schema as an example:

```
// Avro
record Struct {
  int f0;
  int f1;
}

// Parquet
message StructArray {
  required struct_array (LIST) {
    repeated group array {
      required int32 f0;
    }
  }
}
```

According to the condition of this `if` branch, we should interpret `struct_array` as a `LIST` of Avro `Struct`, with all `f1` field set as `null`. However, `f1` defined in `Struct` isn't optional and can't be null.

Did I miss something here?

liancheng replied:
My last comment is a little bit ambiguous. In the example I made, the Avro record `Struct` is the `elementSchema`, while the Parquet repeated group `array` is the `repeatedType` in the code.

liancheng replied:
I guess what is missing here is that we should check all the missing fields are indeed optional, unless this constraint has already been guaranteed somewhere else (then let's document it in the comment).

rdblue replied:
This is for the case where only one field of a nested record is projected. Avro has a projection schema and an expected schema so they can differ. Previously, the logic to match the requested schema (elementSchema) with the file's structure checked whether the name of the single field matched the name of the single field of the element schema. But that left out the case where the two do match (as in your example) but the requested schema has an optional field that should be defaulted.

Whether the other field is optional doesn't matter at this point in the code. Avro fills in default values later and throws an exception if something is missing and has no default at that point.

liancheng replied:
Oh I see, this totally makes sense. Thanks for the explanation!

liancheng commented on parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConverter.java:
```diff
@@ -98,6 +103,39 @@ private static StructType toStructType(TStructDescriptor struct) {
+      }
+      // If the repeated type is a subset of the structure of the ThriftField,
+      // then it must be the element type.
+      return fieldNames.contains(repeatedType.asGroupType().getFieldName(0));
```
Same comments as the ones I made in `AvroRecordConverter.java`.

rdblue replied:
Same as above. The field projection (set by a filter expression) is independent of the Thrift type used to carry data. Nulls are added later and handled separately.

liancheng commented:
Got some questions about `isListElementType` updates.

Also, it would be nice to add parquet-thrift test cases that cover PARQUET-364, e.g., something like [this][1].

[1]: https://github.com/liancheng/parquet-compat/blob/with-parquet-mr-1.7.0/src/test/scala/com/databricks/parquet/schema/ListCompatibilitySuite.scala#L62-L111

rdblue commented:
@liancheng, you're right that I still need to add more test cases, thanks for pointing that out.

rdblue commented:
@liancheng, I added all of the test cases that were added in PARQUET-364 to this commit. I think that addresses all of your concerns.

@isnotinvain, do you still want to have a look at this?

liancheng commented:
+1

rdblue commented:
Thanks @liancheng!
